src/lobster/callbacks/_perturbation_analysis_callback.py

```diff
@@ -0,0 +1,353 @@
+        return embeddings
+
+    def _compute_shuffling_distances(self, model: L.LightningModule, sequence: str) -> list[float]:
+        """Compute distances between original and shuffled sequence embeddings."""
+        original_embedding = self._embed_sequences(model, [sequence]).squeeze(0)
+        shuffling_distances = []
+
+        for _ in range(self.num_shuffles):
+            # Create shuffled sequence
+            shuffled_chars = list(sequence)
+            random.shuffle(shuffled_chars)
+            shuffled_sequence = "".join(shuffled_chars)
+
+            # Get embedding for shuffled sequence
+            shuffled_embedding = self._embed_sequences(model, [shuffled_sequence]).squeeze(0)
+
+            # Compute cosine distance
+            distance = cosine(original_embedding.cpu().numpy(), shuffled_embedding.cpu().numpy())
+            shuffling_distances.append(distance)
+
+        return shuffling_distances
+
+    def _compute_mutation_distances(self, model: L.LightningModule, sequence: str) -> np.ndarray:
+        """Compute distances between original and single-point mutation embeddings."""
+        original_embedding = self._embed_sequences(model, [sequence]).squeeze(0)
+        mutation_distances = []
+
+        for i in range(len(sequence)):
+            position_distances = []
+            for token in self.mutation_tokens:
+                # Create mutated sequence
+                mutated_sequence = sequence[:i] + token + sequence[i + 1 :]
+
+                # Get embedding for mutated sequence
+                mutated_embedding = self._embed_sequences(model, [mutated_sequence]).squeeze(0)
+
+                # Compute cosine distance
+                distance = cosine(original_embedding.cpu().numpy(), mutated_embedding.cpu().numpy())
+                position_distances.append(distance)
+
+            mutation_distances.append(position_distances)
+
+        return np.array(mutation_distances)
+
+    def _create_perturbation_heatmap(self, mutation_distances: np.ndarray, sequence: str, output_file: UPath) -> None:
+        """Create and save a heatmap of mutation perturbations."""
+        df_perturbations = pd.DataFrame(
+            mutation_distances,
+            index=[sequence[i] for i in range(mutation_distances.shape[0])],
+            columns=self.mutation_tokens,
+        )
+
+        plt.figure(figsize=(12, 10))
+        sns.heatmap(
+            df_perturbations.T,
+            cmap="viridis",
+            cbar_kws={"label": "Cosine Distance"},
+            vmax=df_perturbations.values.max(),
+            vmin=df_perturbations.values.min(),
+        )
+
+        plt.xlabel("Sequence Position")
+        plt.ylabel(f"{self.modality.replace('_', ' ').title()} Token")
+        plt.title(f"{self.modality.replace('_', ' ').title()} Mutation Perturbation Analysis")
+        plt.tight_layout()
+        plt.savefig(output_file, dpi=300, bbox_inches="tight")
+        plt.close()
+
+    def _run_analysis(
```

Comment:
can we factor out the actual analysis / metrics calculation code into `lobster.metrics` so it can be easily run any time?